src/librustc/infer/mod.rs

```diff
@@ -181,6 +181,9 @@ pub struct InferCtxt<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     // obligations within. This is expected to be done 'late enough'
     // that all type inference variables have been bound and so forth.
     region_obligations: RefCell<Vec<(ast::NodeId, RegionObligation<'tcx>)>>,
+
+    // true if trait selection in this context should emit `default impl` candiates
+    pub emit_defaul_impl_candidates: Cell<bool>,
```

Comment:
Sorry again for the delay. This flag doesn't feel right. It seems like it's true when we are checking the default impl itself, is that right? Can you give an example of the errors you were getting that you wanted to solve via this flag? 

My guess is that the right way to solve this would be by adding something into the parameter environment for a default impl; for example, an assumption that `Self: Trait<..>`, rather like we have [the trait definition itself](https://github.com/rust-lang/rust/blob/abe85ab0b232e744b602d20a65fbe92aa71c6045/src/librustc_typeck/collect.rs#L1436-L1439), which is -- after all -- a kind of default impl.

Reply:
Oh, I see you left a pretty detailed comment saying why you added it. =) That sounds about like what I was expecting -- but I don't think the flag is the right way to go. We would rather approach this through the parameter environment, I think.